File: src/main/java/org/teamvoided/transition/api/misc/MapCodecs.java
package org.teamvoided.transition.api.misc;

import com.mojang.datafixers.util.Pair;
import com.mojang.serialization.Codec;

import java.util.Map;

public interface MapCodecs {
    static <F, S> Codec<Map<F, S>> codec(Codec<Pair<F, S>> pair) {
        return pair.listOf().xmap(
                list -> list.stream().collect(Pair.toMap()),
                map -> map.entrySet().stream().map(entry -> Pair.of(entry.getKey(), entry.getValue())).toList()
        );
    }
}

File: src/main/kotlin/org/teamvoided/transition/api/misc/MapCodecs.kt
package org.teamvoided.transition.api.misc

import com.mojang.datafixers.util.Pair
import com.mojang.serialization.Codec

object MapCodecs {
    fun <F, S> codec(pair: Codec<Pair<F, S>>): Codec<MutableMap<F, S>> {
        return pair.listOf().xmap({ it.stream().collect(Pair.toMap()) })
        { map -> map.entries.stream().map { Pair.of(it.key, it.value) }.toList() }
    }
}
